File: js/2961.0_Double_Modular_Exponentiation.js
/**
 * @param {number[][]} variables
 * @param {number} target
 * @return {number[]}
 */
var getGoodIndices = function(variables, target) {
    const ans = [];
    for (let i = 0; i < variables.length; i++) {
        const v = variables[i];
            if (powMod(powMod(v[0], v[1], 10), v[2], v[3]) == target) {
                ans.push(i);
            }
    }
    return ans;
};

function powMod(x, y, mod) {
    let res = 1;
    while (y > 0) {
        if ((y & 1) == 1) {
            res = res * x % mod;
        }
        x = x * x % mod;
        y >>= 1;
    }
    return res;
}

File: leetcode-cn/2961.0_Double_Modular_Exponentiation.py
'''
57ms 击败8.51%
16.35MB 击败91.49%
'''

class Solution:
    def getGoodIndices(self, variables: List[List[int]], target: int) -> List[int]:
        ans = []
        for i, variable in enumerate(variables):
            a, b, c, m = variable
            x = pow(a, b, 10)
            y = pow(x, c, m)
            if y == target:
                ans.append(i)
        return ans


class Solution:
    def getGoodIndices(self, variables: List[List[int]], target: int) -> List[int]:
        return [j for j, (a, b, c, m) in enumerate(variables) if pow(pow(a, b, 10), c, m) == target]


class Solution:
    def getGoodIndices(self, variables: List[List[int]], target: int) -> List[int]:
        def pow_mod(x: int, y: int, mod: int) -> int:
            res = 1
            while y:
                if y & 1:
                    res = res * x % mod 
                x = x * x % mod 
                y >>= 1
            return res 

        return [j for j, (a, b, c, m) in enumerate(variables) if pow_mod(pow_mod(a, b, 10), c, m) == target]

